logic/logic.go

```diff
@@ -29,16 +32,58 @@ type operation struct {
 	Group         []string
 }
 
+// Deprecated
 var spec *langSpec
+
+// Deprecated
 var opcodes []operation
 
 // CheckProgram performs basic program validation: instruction count and program cost
+// Deprecated: `langspec.json` can no longer correctly to depicting the cost model (as of 2022.08.22),
+// also to minimize the work in updating SDKs per AVM release, we are deprecating`langspec.json` across all SDKs.
+// The behavior of `CheckProgram` relies on `langspec.json`. Thus, this method is being deprecated.
 func CheckProgram(program []byte, args [][]byte) error {
 	_, _, err := ReadProgram(program, args)
 	return err
 }
 
+func isAsciiPrintableByte(symbol byte) bool {
+	isBreakLine := symbol == '\n'
+	isStdPrintable := symbol >= ' ' && symbol <= '~'
+	return isBreakLine || isStdPrintable
+}
+
+func isAsciiPrintable(program []byte) bool {
+	for _, b := range program {
+		if !isAsciiPrintableByte(b) {
+			return false
+		}
+	}
+	return true
+}
+
+// SanityCheckProgram performs heuristic program validation:
+// check if passed in bytes are Algorand address or is B64 encoded, rather than Teal bytes
+func SanityCheckProgram(program []byte) error {
+	if len(program) == 0 {
+		return fmt.Errorf("empty program")
+	}
+	if isAsciiPrintable(program) {
+		if _, err := types.DecodeAddress(string(program)); err == nil {
+			return fmt.Errorf("requesting program bytes, get Algorand address")
+		}
+		if _, err := base64.StdEncoding.DecodeString(string(program)); err == nil {
+			return fmt.Errorf("program should not be b64 encoded")
+		}
+		return fmt.Errorf("program bytes are all ASCII printable characters, not looking like Teal byte code")
+	}
+	return nil
+}
+
 // ReadProgram is used to validate a program as well as extract found variables
+// Deprecated: `langspec.json` can no longer correctly to depicting the cost model (as of 2022.08.22),
+// also to minimize the work in updating SDKs per AVM release, we are deprecating`langspec.json` across all SDKs.
+// The behavior of `ReadProgram` relies on `langspec.json`. Thus, this method is being deprecated.
```

Comment:
@ahangsu I appreciate the added context here.  I propose these tweaks:
* Avoid using a date - it's hard to correlate a date with a released version.  Either prefer version where deprecation begins (e.g. v1.20.0 for this repo as of writing) or leaving without specification.
* Focus message on langspec.json inaccuracies rather than maintainenance concerns.

To minimize multiple iterations, I also suggest aligning all reviewers on phrasing.  I'm happy to speak live if it'll help us reach a conclusion faster.

```suggestion
// Deprecated: Validation relies on metadata (`langspec.json`) that does not accurately represent opcode behavior across program versions.
// The behavior of `ReadProgram` relies on `langspec.json`. Thus, this method is being deprecated.
```